src/main/java/net/minecraftforge/fml/common/FMLModContainer.java

```diff
@@ -595,16 +593,31 @@ public void constructMod(FMLConstructionEvent event)
     @Subscribe
     public void handleModStateEvent(FMLEvent event)
     {
-        if (!eventMethods.containsKey(event.getClass()))
+        if (!eventHandlers.containsKey(event.getClass()))
         {
             return;
         }
         try
         {
-            for (Method m : eventMethods.get(event.getClass()))
-            {
-                m.invoke(modInstance, event);
-            }
+            eventHandlers.get(event.getClass())
+                    .stream()
+                    .map(eventHandler ->
+                    {
+                        try
+                        {
+                            return eventHandler.handleEvent(event);
+                        }
+                        catch (Throwable t)
+                        {
+                            CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+                            future.completeExceptionally(t);
+                            return future;
+                        }
+                    })
+                    .reduce((future1, future2) -> future1.thenComposeAsync(aVoid -> future2))
+                    .orElse(CompletableFuture.completedFuture(null))
+                    .toCompletableFuture()
```

Comment:
Is this guaranteed to be supported?

Reply:
Yes. Every implementation of CompletionStage MUST implement CompletionStage.toCompletableFuture(), which is meant to return a CompletableFuture that completes when the other CompletionStage implementation does and then holds the same value. So unless the implementation of CompletionStage is bugged this should work.

Reply:
No. Implementations may throw `UnsupportedOperationException`.

Reply:
@diesieben07 Sorry. You're right. I've only read the short description in the Javadoc before. I've now changed that in a way that is guaranteed to always return a CompletableFuture.